File: nextjs-frontend/lib/prefectures.ts
/**
 * Prefecture-related types and utilities
 */

export interface PrefectureStats {
  prefecture: string;
  spot_count: number;
  goshuin_count: number;
}

export interface PrefectureStatsResponse {
  by_prefecture: PrefectureStats[];
  total_prefectures: number;
  total_spots: number;
  total_goshuin: number;
}

/**
 * Japanese hiragana groupings for prefecture organization
 */
export const HIRAGANA_GROUPS = {
  あ: ["愛知県", "青森県", "秋田県", "石川県", "茨城県", "岩手県", "愛媛県", "大阪府", "岡山県", "沖縄県"],
  か: ["香川県", "鹿児島県", "神奈川県", "岐阜県", "京都府", "熊本県", "群馬県", "高知県"],
  さ: ["埼玉県", "佐賀県", "滋賀県", "静岡県", "島根県"],
  た: ["千葉県", "東京都", "栃木県", "鳥取県", "富山県", "徳島県"],
  な: ["長崎県", "長野県", "奈良県", "新潟県"],
  は: ["北海道", "兵庫県", "広島県", "福井県", "福岡県", "福島県"],
  ま: ["三重県", "宮城県", "宮崎県"],
  や: ["山形県", "山口県", "山梨県"],
  わ: ["和歌山県"],
} as const;

export type HiraganaGroup = keyof typeof HIRAGANA_GROUPS;

/**
 * Get the hiragana group for a prefecture
 */
export function getPrefectureGroup(prefecture: string): HiraganaGroup | null {
  for (const [group, prefectures] of Object.entries(HIRAGANA_GROUPS)) {
    if (prefectures.includes(prefecture)) {
      return group as HiraganaGroup;
    }
  }
  return null;
}

/**
 * Group prefectures by hiragana
 */
export function groupPrefecturesByHiragana(
  stats: PrefectureStats[]
): Record<HiraganaGroup, PrefectureStats[]> {
  const grouped: Record<string, PrefectureStats[]> = {
    あ: [],
    か: [],
    さ: [],
    た: [],
    な: [],
    は: [],
    ま: [],
    や: [],
    わ: [],
  };

  for (const stat of stats) {
    const group = getPrefectureGroup(stat.prefecture);
    if (group) {
      grouped[group].push(stat);
    }
  }

  return grouped as Record<HiraganaGroup, PrefectureStats[]>;
}

File: fastapi_backend/app/api/routes/prefectures.py
"""Routes for prefecture statistics."""

from __future__ import annotations

from fastapi import APIRouter
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import CurrentUser, DatabaseSession
from app.models import GoshuinRecord, Spot, User
from app.schemas import PrefectureStats, PrefectureStatsResponse

router = APIRouter(tags=["prefectures"])


@router.get("/stats", response_model=PrefectureStatsResponse)
async def get_prefecture_statistics(
    db: DatabaseSession,
    user: CurrentUser,
) -> PrefectureStatsResponse:
    """Get aggregated statistics grouped by prefecture for the current user.

    Returns spot counts and goshuin record counts for each prefecture
    where the user has activity.
    """
    # Query to get spot counts by prefecture
    spot_stats_query = (
        select(
            Spot.prefecture,
            func.count(Spot.id).label("spot_count")
        )
        .where(Spot.user_id == user.id)
        .group_by(Spot.prefecture)
    )
    spot_stats_result = await db.execute(spot_stats_query)
    spot_stats = {row.prefecture: row.spot_count for row in spot_stats_result}

    # Query to get goshuin counts by prefecture (via spot relationship)
    goshuin_stats_query = (
        select(
            Spot.prefecture,
            func.count(GoshuinRecord.id).label("goshuin_count")
        )
        .join(GoshuinRecord, GoshuinRecord.spot_id == Spot.id)
        .where(
            Spot.user_id == user.id,
            GoshuinRecord.user_id == user.id
        )
        .group_by(Spot.prefecture)
    )
    goshuin_stats_result = await db.execute(goshuin_stats_query)
    goshuin_stats = {row.prefecture: row.goshuin_count for row in goshuin_stats_result}

    # Combine statistics for all prefectures
    all_prefectures = set(spot_stats.keys()) | set(goshuin_stats.keys())

    prefecture_list = [
        PrefectureStats(
            prefecture=prefecture,
            spot_count=spot_stats.get(prefecture, 0),
            goshuin_count=goshuin_stats.get(prefecture, 0)
        )
        for prefecture in sorted(all_prefectures)
    ]

    # Calculate totals
    total_spots = sum(stats.spot_count for stats in prefecture_list)
    total_goshuin = sum(stats.goshuin_count for stats in prefecture_list)

    return PrefectureStatsResponse(
        by_prefecture=prefecture_list,
        total_prefectures=len(prefecture_list),
        total_spots=total_spots,
        total_goshuin=total_goshuin
    )
